Limit JITServer AOT cache memory usage

cjjdespres opened the pull request:
The record creation methods of the JITServer AOT cache now check that
total record allocations do not exceed a configurable maximum before
creating new records. This should be correlated with the total
JITServer AOT cache memory usage, and so provide a bound on that value.

Fixes: https://github.com/eclipse-openj9/openj9/issues/15480

Signed-off-by: Christian Despres <despresc@ibm.com>

cjjdespres commented:
Attn @mpirvu. This is a preliminary version of the feature, with a new `-XX:+JITServerAOTmx=` option for testing purposes. Would simply running one of the usual benchmarks with `-XX:+JITServerUseAOTCache` and a low AOT maximum be a good way of testing it?

I've looked at the usage of the `get*Record` functions that I changed and noticed one or two places where NULL checking is required, but for some of the other use sites it was less clear whether or not it would be a problem.

mpirvu commented on runtime/compiler/env/J9PersistentInfo.hpp:
```diff
@@ -169,6 +169,7 @@ class PersistentInfo : public OMR::PersistentInfoConnector
          _requireJITServer(false),
          _localSyncCompiles(true),
          _JITServerAOTCacheName(),
+         _JITServerAOTMaxBytes(300000000), // 300M
```
I suggest using something like 300*1024*1024 to avoid counting zeros and prevent errors.

mpirvu commented on runtime/compiler/runtime/JITServerAOTCache.cpp:
```diff
@@ -702,6 +720,14 @@ JITServerAOTCache::printStats(FILE *f) const
+   {
+   // The AOT cache record allocations are used as a stand-in for the total memory used by all AOT caches
+   auto aotTotalRecordAllocations = TR::Compiler->persistentGlobalMemory()->_totalPersistentAllocations[TR_Memory::JITServerAOTCache];
+   auto aotMaxBytes = TR::CompilationInfo::get()->getPersistentInfo()->getJITServerAOTMaxBytes();
```
Maybe we should store _JITServerAOTMaxBytes directly into  `JITServerAOTCache` as a static. 
I am also leaning towards making`cacheHasSpace()` a static method to reinforce the idea that this query refers to all the AOT cache, not just to the current cache object.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -778,7 +778,15 @@ ClientSessionData::getOrCreateAOTCache(JITServer::ServerStream *stream)
+            }
+         else
+            {
+            _vmInfo->_useAOTCache = false;
```
I think we need a message in the verbose log for this condition.

mpirvu reviewed (commented):
Overall, it looks good.I have some inline comments.

mpirvu commented:
> Would simply running one of the usual benchmarks with -XX:+JITServerUseAOTCache and a low AOT maximum be a good way of testing it?

Yes, though this would be a manual test, with several clients (maybe differently named caches as well) , and watching that the AOT cache memory does not increase. We should also print a message into the server vlog the first time we reach the AOT cache limit.

mpirvu commented:
The description of the PR should include a reference to the corresponding issue.

mpirvu commented:
> I've looked at the usage of the get*Record functions that I changed and noticed one or two places where NULL checking is required,

Have you already made the required changes? I saw one modification.

>  but for some of the other use sites it was less clear whether or not it would be a problem.

I have yet to check all the places.

cjjdespres commented:
I've made the modifications that I can see need to be made. For some of the other uses (like `getClassChainRecord`) there seemed to be later checking in place that would count a NULL result as a permanent cache miss and disable AOT storing/loading for that method, so it didn't seem like modifications needed to be made. In other cases the result of a `get*Record` would be stored in a struct somewhere, and it was unclear exactly how it was subsequently used.

cjjdespres commented:
Other than addressing the review comments, I've also tested the changes with the acmeair benchmark and a low AOT cache maximum. Judging from the verbose log, new AOT method records do seem to be prevented from being allocated once the threshold is reached, though there are a reasonable amount of errors in the vlog from

https://github.com/eclipse-openj9/openj9/blob/ad49776ddb617a668b4d26eee868ac156baf1d09/runtime/compiler/control/JITServerCompilationThread.cpp#L882-L886

I could disable that log message if the cache is full, perhaps.

AlexeyKhrabrov reviewed (changes requested):
Once the cache size limit is reached, new clients should not be able to create new AOT cache instances. `JITServerAOTCacheMap::get()` should return `NULL` in such case, and this needs to be handled with an error message to vlog in `JITClientSession.cpp`. Also see some minor comments inline.

I'll need to look through more of the existing code, namely `J9AheadOfTimeCompile`, to confirm that these changes are sufficient for the "read only mode" to function correctly.

AlexeyKhrabrov commented:
> though there are a reasonable amount of errors in the vlog from
> 
> https://github.com/eclipse-openj9/openj9/blob/ad49776ddb617a668b4d26eee868ac156baf1d09/runtime/compiler/control/JITServerCompilationThread.cpp#L882-L886
> 
> I could disable that log message if the cache is full, perhaps.

This message should not be disabled if the cache is full. There are two distinct error cases now:
1. Failure to allocate the class chain record (or any of its underlying class or class loader records) due to the cache size limit.
2. Failure to create the class chain caused by e.g. a failure to identify a class loader at the client, which is what the existing error message is about.

Ideally we should to distinguish between these two cases, but that might require more extensive changes to the code, I'm not sure if it's worth it.

mpirvu commented on runtime/compiler/runtime/JITServerAOTCache.cpp:
```diff
@@ -26,6 +26,9 @@
 #include "runtime/JITServerAOTCache.hpp"
 #include "runtime/JITServerSharedROMClassCache.hpp"
 
+size_t JITServerAOTCacheMap::_cacheMaxBytes = 300*1024*2014;
```
The default SCC size is 300 MB. @cjjdespres could you please determine how much AOT cache space is consumed by something like Acmeair?
Perhaps the AOT cache limit should also be a factor of the memory limit for the jitserver process.

AlexeyKhrabrov replied:
I remembered that I actually did these measurements already for the JITServer ATC paper, and the largest AOT cache size (measured the same way as here) I saw was ~130 MB for DayTrader7. Although this number does slowly grow over time as the "tail" of methods that rarely get compiled (and rarely get reused) accumulates in the cache. I think 300 MB is reasonable, but then ideally it should be the same as the default SCC size rather than a separate hard-coded constant. It's important that this number is per application, so we need to be very clear in the documentation that if multiple distinct applications use the same JITServer instance, the limit may need to be increased.

cjjdespres replied:
I saw around 55MB with the AcmeAir benchmark after a few rounds of client connections and loads, but that's a lighter benchmark than DT7. Disabling the limit by default might be better, to keep the behaviour consistent with what it currently is.

mpirvu replied:
I am ok with having no limit by default, given that people typically set a limit for the entire process. What happens if a memory allocation for AOT cache data structures fails? Do we handle that case gracefully?

AlexeyKhrabrov replied:
Without the limit, AOT cache allocation failures are currently handled as `std::bad_alloc` exceptions which IIRC result in failed compilations. I think the same applies to everything else using the persistent allocator, i.e. client session caches, shared ROMClass cache, persistent IProfiler and CHTable data, etc. The only thing that is different about the AOT cache is that it always persists beyond client sessions, so it can grow unbounded.

I don't think that keeping existing behaviour is a concern at this point, thus setting a generous default limit is a good solution for now, but ideally we need to evict stale records and method from the AOT cache. I think the default SCC size (300 MB) is the minimum value that could be considered "generous". I don't think we should go above 2x of that. Also in my experience other memory usage (scratch and client caches) dominates the JITServer memory footprint with multiple concurrent clients anyway, which is where the AOT cache is most beneficial.

cjjdespres replied:
Currently a `std::bad_alloc()` is thrown, which I think is caught in at least `TR::CompilationInfoPerThreadRemote::processEntry`.

mpirvu commented on runtime/compiler/control/J9Options.cpp:
```diff
@@ -1227,6 +1229,16 @@ static bool JITServerParseCommonOptions(J9JavaVM *vm, TR::CompilationInfo *compI
+   if (xxJITServerAOTmxArgIndex >= 0)
+      {
+      uint32_t aotMaxBytes = 0;
+      IDATA ret = GET_INTEGER_VALUE(xxJITServerAOTmxArgIndex, xxJITServerAOTmxOption, aotMaxBytes);
```
We should use `GET_MEMORY_VALUE` instead of `GET_INTEGER_VALUE` because the former allows expressions like 300m.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -778,7 +778,19 @@ ClientSessionData::getOrCreateAOTCache(JITServer::ServerStream *stream)
       if (auto aotCacheMap = TR::CompilationInfo::get()->getJITServerAOTCacheMap())
          {
          _aotCache = aotCacheMap->get(_aotCacheName, _clientUID);
```
This `get` operation creates a new empty AOT cache with the given name (if it does not exist). Maybe we should test whether we reached the limit and refuse to create one if we did. Of course, we need a test for `_aotCache != NULL` for the next operation.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -775,10 +775,34 @@ ClientSessionData::getOrCreateAOTCache(JITServer::ServerStream *stream)
+         _vmInfo->_useAOTCache = false;
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+               "clientUID=%llu requested AOT cache but the the AOT cache size limit has been reached, disabling AOT cache",
```
"but the the AOT" --> "but the AOT"

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -778,7 +778,31 @@ ClientSessionData::getOrCreateAOTCache(JITServer::ServerStream *stream)
+            }
+         else
+            {
+            _vmInfo->_useAOTCache = false;
```
Need to set `_aotCache` to `NULL` as well, otherwise this client may still try to use it in some cases.

AlexeyKhrabrov commented on runtime/compiler/control/J9Options.cpp:
```diff
@@ -1235,6 +1237,15 @@ static bool JITServerParseCommonOptions(J9JavaVM *vm, TR::CompilationInfo *compI
+   if (xxJITServerAOTmxArgIndex >= 0)
+      {
+      uint32_t aotMaxBytes = 0;
+      if (GET_INTEGER_VALUE(xxJITServerAOTmxArgIndex, xxJITServerAOTmxOption, aotMaxBytes) == OPTION_OK)
```
This is still using `GET_INTEGER_VALUE` instead of `GET_MEMORY_VALUE`, which requires specifying the limit in bytes if I understand correctly.

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk17

AlexeyKhrabrov commented:
Current implementation does not correctly handle some of the cases when the limit is reached in the middle of a compilation.

For instance, a NULL check is necessary here to prevent a segfault:
https://github.com/cjjdespres/openj9/blob/5c7fc255094a475103a7fc22bf3c264479e1f90c/runtime/compiler/runtime/JITClientSession.cpp#L834.

The logic for requesting missing information from the client in `ClientSessionData::getClassRecord(J9Class *, JITServer::ServerStream *)` and `ClientSessionData::getClassChainRecord()` relies on the NULL return values and the `missingLoaderRecord` flag set here: https://github.com/cjjdespres/openj9/blob/5c7fc255094a475103a7fc22bf3c264479e1f90c/runtime/compiler/runtime/JITClientSession.cpp#L856. Since this flag is set to true in case of reaching the size limit, this will result in needless client messages. We need to properly distinguish the "cache size limit reached" case from "missing class loader info" and "uncached class info". Unfortunately I didn't consider limiting the cache size when I wrote this code.

The rest of the code (outside of `JITClientSession.cpp`) seems to correctly handle NULLs returned from `ClientSessionData::get*Record()` methods, but you should double check that.

cjjdespres commented:
The other `get*Record` invocations still seem okay to me.

mpirvu commented:
A suggestion:  The implementation could become simpler if we check cache limits only when a compilation starts and sets `_aotCacheStore`. If we find there is cache space and decided to set `_aotCacheStore = true;` then we can allow the compilation to allocate as much memory as it needs, despite exceeding the limit. This of course assumes that a single compilation cannot allocate a humongous amount of memory for the  AOT cache.

AlexeyKhrabrov commented:
The real issue is with loads, not compilations. Loads must be allowed to proceed as long as they don't create new records. If a loads tries to create a new record, we need to immediately fail it (it's guaranteed to fail eventually anyway), otherwise we can still have unbounded growth. And since there are other "errors" we already handle on this path (uncached classes, missing class loader info), we need to carefully handle both those and AOT cache allocation failures. So we already need to do this for class loader, class, and class chain records on the load path, and the rest of the code (on the compilation path) already handles failures to create records.

Although there might be a simpler way to handle this using exceptions rather than manually propagating error information up the call stack.

AlexeyKhrabrov commented:
I'm still reviewing the rest of the changes, but one thing I've noticed is that `JITServerAOTCache::storeMethod()` doesn't check the cache size limit, which means that if the limit is reached in the middle of a compilation, but no new records need to be created until we store the compiled method, the store will succeed, exceeding the limit. I don't think that this is an issue since the limit is not accurate anyway, but we might want to keep things consistent and fail the store (with a message to vlog). I'm fine either way, but if we keep the current behaviour, I think we need a comment about this in `JITServerAOTCache::storeMethod()`.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.hpp:
```diff
@@ -460,19 +460,28 @@ class ClientSessionData
+   const AOTCacheClassRecord *getClassRecord(ClientSessionData::ClassInfo &classInfo, bool &missingLoaderInfo);
+   // If this function sets one of the two boolean flags then a NULL result is due to one of those error conditions; otherwise
+   // that result is due to a failure to allocate.
+   const AOTCacheClassRecord *getClassRecord(J9Class *clazz, bool &missingLoaderRecord, bool &uncachedClass);
```
Should be `bool &missingLoaderInfo` to be consistent with other functions.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -819,27 +849,32 @@ ClientSessionData::getClassRecord(ClientSessionData::ClassInfo &classInfo)
+      }
 
-   auto record = getClassRecord(it->second);
+   bool missingLoaderInfo = false;
```
Should simply pass down the reference to the flag passed as argument (which should be renamed), otherwise we lose this error information here.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -819,27 +849,32 @@ ClientSessionData::getClassRecord(ClientSessionData::ClassInfo &classInfo)
-      record = getClassRecord(clazz, missingLoaderRecord);
+      record = getClassRecord(clazz, missingLoaderRecord, uncachedClass);
       }
 
```
If the `getClassRecord()` call above failed because we reached the cache size limit, it makes more sense to fail immediately so that we don't communicate with the client unnecessarily.

cjjdespres replied:
Isn't that what happens already? If `getClassRecord()` fails because the cache size limit was reached then both `missingLoaderInfo` and `uncachedClass` will be false.

AlexeyKhrabrov replied:
Right, I missed that. Maybe you should add a comment.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -857,7 +892,7 @@ ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *strea
          it->second._classChainOffsetIdentifyingLoader = offset;
          it->second._classNameIdentifyingLoader = name;
-         record = getClassRecord(it->second);
+         record = getClassRecord(it->second, missingLoaderInfo);
```
Can add a `TR_ASSERT` that `missingLoaderInfo` is false because we've just received it from the client.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -867,15 +902,15 @@ ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *strea
          );
```
Losing error information here because `missingLoaderInfo` was never set to true.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -819,27 +849,32 @@ ClientSessionData::getClassRecord(ClientSessionData::ClassInfo &classInfo)
+ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *stream, bool &missingLoaderInfo)
    {
    const AOTCacheClassRecord *record = NULL;
    bool missingLoaderRecord = false;
```
Should simply pass down `missingLoaderInfo`. It will stay set to true if we request the missing info from the client and still fail.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -867,15 +902,15 @@ ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *strea
       JITServerHelpers::cacheRemoteROMClassOrFreeIt(this, clazz, romClass, classInfoTuple);
 
       OMR::CriticalSection cs(getROMMapMonitor());
-      record = getClassRecord(clazz, missingLoaderRecord);
```
Should pass down `missingLoaderInfo` - it can still be set to true. Can also `TR_ASSERT` that `uncachedClass` is false.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -957,10 +1000,18 @@ ClientSessionData::getClassChainRecord(J9Class *clazz, uintptr_t *classChain,
+               }
+            else
+               {
+               return NULL;
```
Should add a comment that this is an allocation failure.

cjjdespres commented:
I think it's reasonable to include that check.

AlexeyKhrabrov reviewed (changes requested):
Seems like there are still some issues in handling the different error cases in `ClientSessionData::getClassRecord()`.

cjjdespres commented:
I haven't yet marked those two fields as volatile, but otherwise the remaining changes should be finished.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -857,7 +889,8 @@ ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *strea
          it->second._classNameIdentifyingLoader = name;
-         record = getClassRecord(it->second);
+         record = getClassRecord(it->second, missingLoaderInfo);
+         TR_ASSERT(!missingLoaderInfo, "Class %p cannot have missing loader info as we've just received it from the client");
```
Missing argument for `%p`.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -867,15 +900,15 @@ ClientSessionData::getClassRecord(J9Class *clazz, JITServer::ServerStream *strea
 
       OMR::CriticalSection cs(getROMMapMonitor());
-      record = getClassRecord(clazz, missingLoaderRecord);
+      record = getClassRecord(clazz, missingLoaderInfo, uncachedClass);
```
I think it's worth adding an assertion that the class cannot be uncached anymore.

AlexeyKhrabrov reviewed (commented):
Other than minor issues with assertions (see inline comments) and making sure that the C++ compiler doesn't optimize/reorder the stores in `getOrCreateAOTCache()` with `volatile` and a barrier, everything else looks good.

cjjdespres commented:
Just to clarify, since I'm not that familiar with `volatile`, you're suggesting

```
   volatile JITServerAOTCache *_aotCache;
   const volatile AOTCacheAOTHeaderRecord *_aotHeaderRecord;
```

in `ClientSessionData`? Is it safe to cast `_aotCache` and `aotHeaderRecord` to non-volatile pointers where they are currently used (except for where we assign to them, of course)?

AlexeyKhrabrov commented:
> ```
>    volatile JITServerAOTCache *_aotCache;
>    const volatile AOTCacheAOTHeaderRecord *_aotHeaderRecord;
> ```

No, this would make the `JITServerAOTCache` and `AOTCacheHeaderRecord` objects themselves `volatile`, not the pointers to them (it's parsed the same way as `const`). We need to make the pointer fields `volatile` to guarantee that they are written to exactly as in the source code (regardless of optimizations). So it should be:
```
JITServerAOTCache *volatile _aotCache;
const AOTCacheAOTHeaderRecord *volatile _aotHeaderRecord;
```

Depending on the platform, that might not be enough, we also need a write barrier (`VM_AtomicSupport::writeBarrier()`) between the stores to `_aotHeaderRecord` and `_aotCache`. Even though triggering this potential race is extremely unlikely in practice, it's not impossible unless we make these changes.

> Is it safe to cast `_aotCache` and `aotHeaderRecord` to non-volatile pointers where they are currently used (except for where we assign to them, of course)?

Yes, and it should be done by the compiler implicitly, you shouldn't need to change any other code.

cjjdespres commented:
I understand. That makes more sense. Yes, the explicit casts would only be needed with my (incorrect) definition.

AlexeyKhrabrov commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -777,8 +777,34 @@ ClientSessionData::getOrCreateAOTCache(JITServer::ServerStream *stream)
+         if (record)
+            {
+            _aotHeaderRecord = record;
+            VM_AtomicSupport::writeBarrier();
```
Could you please add a comment explaining why we need the barrier?

AlexeyKhrabrov reviewed (approved):
LGTM

mpirvu commented on runtime/compiler/control/JITServerCompilationThread.cpp:
```diff
@@ -876,14 +876,18 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
                   "method %p won't be loaded from or stored in AOT cache",
-                  (unsigned long long)clientId, clazz, ramMethod
+                  (unsigned long long)clientId,
+                  missingLoaderInfo ? "missing class loader info" : "the AOT cache size limit",
```
Parameters 2 and 3 need to be switched here.

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk17
